src/mystery_gift.c

```diff
@@ -0,0 +1,283 @@
+        if (!MysteryGiftTagIsValid(mg->cards[i].tag)) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+BOOL SaveMysteryGift_TagAtIndexIsValid(const MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_MYSTERY_GIFTS);
+
+    return MysteryGiftTagIsValid(mg->gifts[index].tag) != FALSE;
+}
+
+BOOL SaveMysteryGift_CardTagIsValid(const MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_WONDER_CARDS);
+
+    return MysteryGiftTagIsValid(mg->cards[index].tag) != FALSE;
+}
+
+BOOL SaveMysteryGift_SpecialCardTagIsValid(const MysteryGiftSave* mg) {
+    return MysteryGiftTagIsValid(mg->specialWonderCard.tag) != FALSE;
+}
+
+BOOL SaveMysteryGift_HasAnyCard(const MysteryGiftSave* mg) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_WONDER_CARDS; ++i) {
+        if (SaveMysteryGift_CardTagIsValid(mg, i) == TRUE) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+BOOL SaveMysteryGift_HasAnyGift(const MysteryGiftSave* mg, int a1) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (MysteryGiftTagIsValid(mg->gifts[i].tag) && a1 == mg->gifts[i].flag) {
+            return TRUE;
+        }
+    }
+
+    return FALSE;
+}
+
+void SaveMysteryGift_SetReceivedByCardId(MysteryGiftSave* mg, int a1) {
+    int i;
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (MysteryGiftTagIsValid(mg->gifts[i].tag) && a1 == mg->gifts[i].flag) {
+            SaveMysteryGift_SetReceivedByIdx(mg, i);
+            break;
+        }
+    }
+}
+
+BOOL SaveMysteryGift_ReceivedFlagTest(const MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_MYSTERY_GIFT_RECV_FLAGS);
+    return mg->receivedFlags[index / 8] & (1 << (index & 7)) ? TRUE : FALSE;
+}
+
+void SaveMysteryGift_ReceivedFlagSet(MysteryGiftSave* mg, int index) {
+    GF_ASSERT(index < NUM_MYSTERY_GIFT_RECV_FLAGS);
+    mg->receivedFlags[index / 8] |= (1 << (index & 7));
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+void SaveMysteryGift_ReceivedFlagClear(MysteryGiftSave* mg, int index) {
+    u8 mask = ~(1 << (index & 7));
+    GF_ASSERT(index < NUM_MYSTERY_GIFT_RECV_FLAGS);
+    mg->receivedFlags[index / 8] &= mask;
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+BOOL SaveMysteryGift_TestFlagx7FF(const MysteryGiftSave* mg) {
+    return SaveMysteryGift_ReceivedFlagTest(mg, 0x7FF);
+}
+
+void SaveMysteryGift_SetFlagx7FF(MysteryGiftSave* mg) {
+    SaveMysteryGift_ReceivedFlagSet(mg, 0x7FF);
+}
+
+// ======================================
+
+static MysteryGiftSave* sMysteryGiftSaveData;
+
+void GetStaticPointerToSaveMysteryGift(SaveData* saveData) {
+    if (sMysteryGiftSaveData == NULL) {
+        sMysteryGiftSaveData = Save_MysteryGift_Get(saveData);
+    }
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+void DeleteStaticPointerToMysteryGift(void) {
+    if (sMysteryGiftSaveData != NULL) {
+        sMysteryGiftSaveData = NULL;
+    }
+}
+
+int GetFirstQueuedMysteryGiftIdx(void) {
```

Comment:
Arguably using the SaveMysteryGift(Data) prefix here is fine since it's modifying the static instance of SaveMysteryGiftData, ie analogous to a static function of the class. Ditto for everything else.

Reply:
In my scrcmd_11 PR, I explore the possibility of this static pointer, at one point in development, holding a transaction. Upon further study of retsam, this was not the case, but wouldn't it be cool if it was?